allow installation of extensions via package managers

boneskull opened the pull request:
## Summary

This enables installation of extensions via `npm`; e.g., `npm install appium-xcuitest-driver -D`.  Because `appium` usually manages the `extensions.yaml` (the "manifest"), this is not possible when installing modules via other means.  To handle this situation, if we have an `appium`-dep-containing `package.json` in `APPIUM_HOME`, a hash is taken of the `package.json`.  This hash is checked at runtime; if it has changed, `appium` will scan `node_modules` (_not_ `package.json`; we only care about extensions that are _actually installed_, which is different from how we handle an `appium` dep) for extensions, and the hash will be updated.  Extensions are defined as "modules having a `package.json` which contains an `appium` field which is an object of a particular shape".  When these are found, they are appended to `extensions.yaml`.  In theory, this allows a user to use `appium` _or_ `npm` (at their convenience) to manage extensions.

See the "when appium is a dependency" suite in `test/cli/cli-e2e-specs.js` for more context.

## BREAKING CHANGES

(potentially, anyway)

1. Installation of extensions via `appium` now treats `$APPIUM_HOME` _mostly_ like a project containing a `node_modules` subfolder, with extensions under that.  Previously, each extension would have its own folder, with a `node_modules` and empty `package.json`, and finally the extension.  The intent was to inhibit deduping and conflicts with any `package.json` files in ancestors of `$APPIUM_HOME`.  I showed this was unnecessary because the same thing can be accomplished via the flag `--global-style` for `npm install`.  Ultimately, this simplifies the directory structure of `$APPIUM_HOME`.
2. `$APPIUM_HOME/extensions.yaml` now lives at `$APPIUM_HOME/node_modules/.cache/appium/extensions.yaml`.  The `node_modules/.cache/<pkg-name>` convention is fairly common for packages having temporary files which need to persist while a module is installed (no more, no less).
3. If the _current working directory_ contains a `package.json` which has `appium` as any sort of (non-`file:`) dependency _or_ `$CWD/node_modules/appium` is a resolvable module, `APPIUM_HOME` defaults to `cwd`.
4. "Install" and "link" lockfiles (see `lib/cli/npm`) have also moved from `$APPIUM_HOME` to `$APPIUM_HOME/node_modules/.cache/appium`.
5. The `runExtensionCommand()` function in the `cli/extension` module no longer accepts a `type` parameter, which can be derived from the `configObject` parameter anyway (redundant)
6. Loading of extensions (i.e. reading `extensions.yaml`) has been moved into its own function, `loadExtensions()` in `extension/index.js`.
7. `getActivePlugins` and `getActiveDrivers` have been moved here from `lib/main`.  Changed the signatures of these, since they don't actually need the entire `ParsedArgs` object.
8. `ext-config-io` module, which handles reading/writing `extensions.yaml` has been moved to `extension/manifest.js`.
9. The `installPath` property in `extensions.yaml` has been removed and is unneeded due to 1.
10. `read()` has been removed from `extension-config` and its subclasses.
11. Moved `ReadonlyMap` from `utils` to `schema/schema` module.  Probably so I didn't have to stub it in a test.
12. Moved `rootDir` from `utils` to `config`.
13. Removed `drivers` module; moved `findMatchingDriver` to `DriverConfig#findMatchingDriver`.  Removed `AppiumDriver#_findMatchingDriver`; calls `DriverConfig#findMatchingDriver` directly instead.
14. Removed `plugins` module; its contents now live in `constants`.
15. Dynamic determination of `APPIUM_HOME` now happens in the `env.resolveAppiumHome()` method of `@appium/support` (introduced in a previous commit).  It should not happen anywhere else, because it will be wrong.
16. Because the `lib` dir was becoming rather unwieldy, I moved modules associated with extension loading and config into an `extension` subdir
17. `getParser()` in `cli/parser` is no longer async; the async stuff it was doing moved into `loadExtensions()`.  `getParser()` can be used without calling `loadExtensions()` first (no extensions will be loaded)

## Notes

- Added a `constants` module, which provides a centralized place for various constants.  This makes finding constants less cognitively demanding. 🙂  You may not like this; if you don't, the constants can be moved to more appropriate places.
- Added `extension/package-changed` module which is used to determine if a `package.json` has changed
- Many types are added/changed throughout.  See especially `extension/manifest` for many type definitions involving the shape of `extensions.yaml` and the `appium` prop of extensions' `package.json`.
- Reorganized some tests based on structure of `lib`.
- Split `extension-config-specs.js` into `driver-config-specs.js` and `plugin-config-specs.js`.
- Many refactors in now-fully-typed `cli-e2e-specs`; introduced some more helpers in `test/cli/cli-helpers.js` and used these.  Makes calling out to the `appium` executable from E2E tests a little easier.
- Introduced `test/extension/mocks` for DRY mocks, since many of the unit tests here now use the same ones.
- Consolidated fixtures into `test/fixtures`. Fixtures can always be found via the `resolveFixture()` method of `test/helpers.js`.

## TODO (before this lands)

- Instead of using `$CWD` for `APPIUM_HOME` if `appium` is a dep, we probably want to use `find-pkg-up` or something.  This should be okay, since if we find a `package.json` and `appium` _isn't_ a dep, we will just be using the default `APPIUM_HOME`.
- We _may_ also need to search `package.json` for known extensions, since it's possible to use them standalone.  However, I'm not sure if this use case needs `APPIUM_HOME` at all?

* * *

First, sorry for another huge PR.  

I want to answer those TODO items before moving this out of draft.   If there's something anyone wishes I pull out of this PR, please lmk.

boneskull committed:
chore(gulp-plugins): add test env for babel

When `BABEL_ENV=test` we should get better sourcemaps

boneskull committed:
chore: typescript updates

- Upgrade TS to `next` tag.  It's "safe" to use using bleeding-edge TS here since we don't actually use it to compile anything
- Add a bunch of missing types: `find-root`, `klaw`, `lockfile`, `mv`, `ncp`, `rimraf`

boneskull committed:
fix(fake-driver): create cjs executable entry point

This driver wasn't working properly when used directly

boneskull committed:
feat(support): add env module

- `env` module exposes functions for determining `APPIUM_HOME`, where various files should live, etc., by examining the runtime environment
- fixed some functions in `fs` which unnecessarily used `this` (which prevented dereferencing)
- added `walk()` to `fs` which just exposes `klaw` (since `walkDir` has very limited behavior). this enables `walk()` to be used as an async iterator, which is handy.
- fixed a bunch of types in the `util` module. involved some actual code changes, but the original code only worked by chance.
- removed `strict` option from `util.coerceVersion()`, since the option isn't actually used anywhere.  this should be considered a breaking change if some module somewhere _else_ uses it and we consider it a public API.  I'm pretty sure that's not the case, so.

boneskull committed:
chore: update wallaby & vscode configs

boneskull committed:
feat(appium): allow installation of extensions via `npm`

## BREAKING CHANGES

(potentially, anyway)

1. Installation of extensions via `appium` now treats `$APPIUM_HOME` _mostly_ like a project containing a `node_modules` subfolder, with extensions under that.  Previously, each extension would have its own folder, with a `node_modules` and empty `package.json`, and finally the extension.  The intent was to inhibit deduping and conflicts with any `package.json` files in ancestors of `$APPIUM_HOME`.  I showed this was unnecessary because the same thing can be accomplished via the flag `--global-style` for `npm install`.  Ultimately, this simplifies the directory structure of `$APPIUM_HOME`.
2. `$APPIUM_HOME/extensions.yaml` now lives at `$APPIUM_HOME/node_modules/.cache/appium/extensions.yaml`.  The `node_modules/.cache/<pkg-name>` convention is fairly common for packages having temporary files which need to persist while a module is installed (no more, no less).
3. If the _current working directory_ contains a `package.json` which has `appium` as any sort of (non-`file:`) dependency _or_ `$CWD/node_modules/appium` is a resolvable module, `APPIUM_HOME` defaults to `cwd`.
4. "Install" and "link" lockfiles (see `lib/cli/npm`) have also moved from `$APPIUM_HOME` to `$APPIUM_HOME/node_modules/.cache/appium`.
5. The `runExtensionCommand()` function in the `cli/extension` module no longer accepts a `type` parameter, which can be derived from the `configObject` parameter anyway (redundant)
6. Loading of extensions (i.e. reading `extensions.yaml`) has been moved into its own function, `loadExtensions()` in `extension/index.js`.
7. `getActivePlugins` and `getActiveDrivers` have been moved here from `lib/main`.  Changed the signatures of these, since they don't actually need the entire `ParsedArgs` object.
8. `ext-config-io` module, which handles reading/writing `extensions.yaml` has been moved to `extension/manifest.js`.
9. The `installPath` property in `extensions.yaml` has been removed and is unneeded due to 1.
10. `read()` has been removed from `extension-config` and its subclasses.
11. Moved `ReadonlyMap` from `utils` to `schema/schema` module.  Probably so I didn't have to stub it in a test.
12. Moved `rootDir` from `utils` to `config`.
13. Removed `drivers` module; moved `findMatchingDriver` to `DriverConfig#findMatchingDriver`.  Removed `AppiumDriver#_findMatchingDriver`; calls `DriverConfig#findMatchingDriver` directly instead.
14. Removed `plugins` module; its contents now live in `constants`.
15. Dynamic determination of `APPIUM_HOME` now happens in the `env.resolveAppiumHome()` method of `@appium/support` (introduced in a previous commit).  It should not happen anywhere else, because it will be wrong.
16. Because the `lib` dir was becoming rather unwieldy, I moved modules associated with extension loading and config into an `extension` subdir
17. `getParser()` in `cli/parser` is no longer async; the async stuff it was doing moved into `loadExtensions()`.  `getParser()` can be used without calling `loadExtensions()` first (no extensions will be loaded)

## Summary

This enables installation of extensions via `npm`; e.g., `npm install appium-xcuitest-driver -D`.  Because `appium` usually manages the `extensions.yaml` (the "manifest"), this is not possible when installing modules via other means.  To handle this situation, if we have an `appium`-dep-containing `package.json` in `APPIUM_HOME`, a hash is taken of the `package.json`.  This hash is checked at runtime; if it has changed, `appium` will scan `node_modules` (_not_ `package.json`; we only care about extensions that are _actually installed_, which is different from how we handle an `appium` dep) for extensions, and the hash will be updated.  Extensions are defined as "modules having a `package.json` which contains an `appium` field which is an object of a particular shape".  When these are found, they are appended to `extensions.yaml`.  In theory, this allows a user to use `appium` _or_ `npm` (at their convenience) to manage extensions.

See the "when appium is a dependency" suite in `test/cli/cli-e2e-specs.js` for more context.

## Notes

- Added a `constants` module, which provides a centralized place for various constants.  This makes finding constants less cognitively demanding. 🙂  You may not like this; if you don't, the constants can be moved to more appropriate places.
- Added `extension/package-changed` module which is used to determine if a `package.json` has changed
- Many types are added/changed throughout.  See especially `extension/manifest` for many type definitions involving the shape of `extensions.yaml` and the `appium` prop of extensions' `package.json`.
- Reorganized some tests based on structure of `lib`.
- Split `extension-config-specs.js` into `driver-config-specs.js` and `plugin-config-specs.js`.
- Many refactors in now-fully-typed `cli-e2e-specs`; introduced some more helpers in `test/cli/cli-helpers.js` and used these.  Makes calling out to the `appium` executable from E2E tests a little easier.
- Introduced `test/extension/mocks` for DRY mocks, since many of the unit tests here now use the same ones.
- Consolidated fixtures into `test/fixtures`. Fixtures can always be found via the `resolveFixture()` method of `test/helpers.js`.

## TODO (before this lands)

- Instead of using `$CWD` for `APPIUM_HOME` if `appium` is a dep, we probably want to use `find-pkg-up` or something.  This should be okay, since if we find a `package.json` and `appium` _isn't_ a dep, we will just be using the default `APPIUM_HOME`.
- We _may_ also need to search `package.json` for known extensions, since it's possible to use them standalone.  However, I'm not sure if this use case needs `APPIUM_HOME` at all?

chore: vscode stuff

- change default build task to `dev`, which just transpiles in the background
- updated `test - current file` launch config for better debugging

chore(appium): peer review refactors

- The old class name of `Manifest` used to be `ExtConfigIO`, but I forgot to update these variable names.  `io` is now `manifest`, wherever used
- Created a more clear separation between the `create()` and `getInstance()` static methods of `DriverConfig`/`PluginConfig`.  `create()` should _only_ be used for instantiation, and `getInstance()` should _only_ be used if we have already instantiated the object with an associated `Manifest`. `create()` now _throws_ if the `Manifest` already has an association in place.  The reason this is split into two functions instead of just using `_.memoize()` is because of the options in the second parameter (`extData`, `logFn`), which would require the _same configuration object_ in order to return the already-created instance.  This is not a problem in `Manifest`, where the factory only accepts a single parameter.
- The `_instances` static method of `PluginConfig`/`DriverConfig` is now a `WeakMap`, keyed on a `Manifest` instance instead of the `appiumHome` string. This makes the relationship between them more explicit.
- Fixed some formatting
- Fixed type issues in schema tests, `grid-register` and its tests
- Made the constructor of `Manifest` private, created a memoized static method `getInstance()`; removed `getManifestInstance()` export.
- Refactored `let`-destructuring to the fancy new syntax I learned from @mykola-mokhnach
- Explained the meaning of `CONFIG_DATA_DRIVER_KEY` and `CONFIG_DATA_PLUGIN_KEY` in `manifest` module
- Replaced unnecessary use of `null` with `undefined` in `Manifest` class
- Removed runtime type check from `packageDidChange()`
- Had some fun renaming `ReadonlyMap` in `schema` module

mykola-mokhnach commented on packages/appium/lib/cli/npm.js:
```diff
@@ -169,13 +187,13 @@ export default class NPM {
     // org, so if so, that will get correctly exploded into multiple directories, by path.resolve here
     // (even on Windows!)
-    const pkgJson = path.resolve(pkgDir, 'node_modules', pkgName, 'package.json');
+    const pkgJsonPath = resolveFrom(this.appiumHome, `${pkgName}/package.json`);
```
based on the comment above is it ok to explicitly use slash as a path separator? would that still work on Windows?

mykola-mokhnach replied:
same below

boneskull replied:
so, yes, because `resolveFrom` uses `require.resolve()` under the hood (which, in turn, is used by `require()`) which normalizes paths.  

If I wanted to build a path but _not_ attempt to resolve it as a module, I'd need to use e.g., `path.join()` or `path.resolve()` and pass multiple args without slashes.

mykola-mokhnach commented on packages/appium/lib/cli/parser.js:
```diff
@@ -2,21 +2,16 @@
 import _ from 'lodash';
 import path from 'path';
-import { DRIVER_TYPE, PLUGIN_TYPE } from '../extension-config';
+import { DRIVER_TYPE, PLUGIN_TYPE, SERVER_SUBCOMMAND} from '../constants';
```
missing space after SERVER_SUBCOMMAND ?

mykola-mokhnach commented on packages/appium/lib/cli/parser.js:
```diff
@@ -258,19 +253,16 @@ class ArgParser {
  */
-async function getParser (debug = false) {
-  await B.all([driverConfig.read(), pluginConfig.read()]);
+function getParser (debug) {
```
why the default value has been removed?

boneskull replied:
it was redundant; `getParser()` is the only code which calls `new ArgParser()` and the signature for the `ArgParser` constructor is `constructor(debug = false)`.  So the default gets set there.  This prevents us from needing to maintain two sets of the same default value.

mykola-mokhnach commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,215 @@
+ */
+export class DriverConfig extends ExtensionConfig {
+  /**
+   * A mapping of `APPIUM_HOME` values to {@link DriverConfig} instances.
```
so this assumes we could have multiple appium homes with different configs on the same machine and switch between them?

boneskull replied:
our tests could do this sort of thing, if they don't already.  but we needed a "key" to ensure we're not creating multiple copies of this class, and `APPIUM_HOME` was obvious, but it doesn't need to be that.  It could be an instance of a `Manifest` object, and we could use a `WeakMap`.

mykola-mokhnach commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,215 @@
+   * @param {ExtRecord<DriverType>} exts
+   */
+  validate (exts) {
+    this.knownAutomationNames.clear();
```
is that a side effect?

boneskull replied:
Yes.  Not my fault. 😄

boneskull replied:
I guess we could do `this.knownAutmationNames = new Set()`

boneskull replied:
On second thought, maybe it is my fault.  Just not _recently_ my fault. :D

So, when we validate, we're validating the entire collection of drivers found in `extensions.yaml`.

`knownAutomationNames` tracks if there's a collision in automation names, since there can only be one driver per name.

`validate()` in `ExtensionConfig` calls `getWhateverProblems()`, and _that_ function begins adding automation names to the `Set`.  

So, what we could do to avoid the side effect is to create a new `Set` in `validate()`, then explicitly pass it into each call to `getWhateverProblems()`.  This is more verbose, but probably a better idea.

boneskull replied:
But, since this is a subclass, and `ExtensionConfig#validate` does the work, we break the abstraction by passing something driver-specific into the abstract `getConfigProblems(..)`.

The refactoring required to make this work due to the inheritance at play may not be worth the bother.

boneskull replied:
... that being said, I think it _should_ be refactored.  Just not here.

mykola-mokhnach commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,215 @@
+  /**
+   * Gets an existing instance of {@link DriverConfig} based value of `io.appiumHome`
+   * @param {import('./manifest').Manifest} io - IO object
+   * @returns {DriverConfig}
```
could the result be undefined?

boneskull replied:
No.  If one does not exist, it will create a new one.

mykola-mokhnach replied:
but this particular call does not check if `_instances` mapping contains `io.appiumHome`, does it?

boneskull replied:
As you mentioned here and below, these were confusing.  I rewrote them (pushing in a few), so:

- `DriverConfig.create(manifest, opts)` only creates an instance, and throws if one is already associated with a `Manifest`.
- `DriverConfig.getInstance(manifest)` returns an existing instance if it exists, or `undefined` if it does not.

(same for `PluginConfig`)

The mapping is now a `WeakMap` based on the `Manifest` instance instead of `appiumHome`.

boneskull replied:
`loadExtensions()` now tries `getInstance()` before calling `create()`.

mykola-mokhnach commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,215 @@
+   */
+  findMatchingDriver ({automationName, platformName}) {
+    if (!_.isString(platformName)) {
+      throw new Error('You must include a platformName capability');
```
not sure if we need to throw a generic error here or more specific one from the list of available W3C error codes

boneskull replied:
old code

mykola-mokhnach commented on packages/appium/lib/extension/extension-config.js:
```diff
@@ -0,0 +1,458 @@
+        continue;
+      }
+      // remove this extension from the list since it's not valid
+      delete exts[extName];
```
is it ok to implicitly mutate the argument?

boneskull replied:
old code.  but... since nothing other than `ExtensionConfig` or its subclasses should ever call this function, it's _probably_ OK.

mykola-mokhnach commented on packages/appium/lib/extension/extension-config.js:
```diff
@@ -0,0 +1,458 @@
+  getSchemaProblems (extData, extName) {
+    const problems = [];
+    const {schema: argSchemaPath} = extData;
+    if (ExtensionConfig.extDataHasSchema(extData)) {
```
I prefer to keep such methods on module level and only make them static if we need this data in subclasses

boneskull replied:
noted

mykola-mokhnach commented on packages/appium/lib/extension/extension-config.js:
```diff
@@ -0,0 +1,458 @@
+  log;
+
+  /** @type {import('./manifest').Manifest} */
+  io;
```
is there any reason to call this variable `io`? I cannot really understand what it contains

boneskull replied:
Yes, that is old. I'll change it.

mykola-mokhnach commented on packages/appium/lib/extension/extension-config.js:
```diff
@@ -0,0 +1,458 @@
+   * @param {ExtName<ExtType>} extName
+   * @returns {ExtClass<ExtType>}
+   */
+  require (extName) {
```
mayt this method shadow the original `require` global?

boneskull replied:
Only if you dereference it, I suppose.  old code

mykola-mokhnach commented on packages/appium/lib/extension/extension-config.js:
```diff
@@ -0,0 +1,458 @@
+   * @returns {boolean}
+   */
+  isInstalled (extName) {
+    return _.includes(Object.keys(this.installedExtensions), extName);
```
you could also use _.keys

boneskull replied:
old code

mykola-mokhnach commented on packages/appium/lib/extension/extension-config.js:
```diff
@@ -0,0 +1,458 @@
+   * Returns `true` if a specific {@link ExtData} object has a `schema` prop.
+   * The {@link ExtData} object becomes a {@link ExtDataWithSchema} object.
+   * @template {ExtensionType} ExtType
+   * @param {ExtData<ExtType>} extData
```
is extData optional?

boneskull replied:
No

mykola-mokhnach commented on packages/appium/lib/extension/index.js:
```diff
@@ -0,0 +1,87 @@
+    } catch (err) {
+      log.error(`Could not load plugin '${pluginName}', so it will not be available. Error ` +
+                   `in loading the plugin was: ${err.message}`);
+      log.debug(err.stack);
```
👍

mykola-mokhnach commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,564 @@
+import _ from 'lodash';
+import path from 'path';
+import YAML from 'yaml';
+import {DRIVER_TYPE, PLUGIN_TYPE} from '../constants';
```
we probably need to follow one style here - either we keep spaced after and before object curly brackets or we don't

boneskull replied:
right, so, the reason this _sometimes_ happens is that we don't seem to have anything that automates the spaces around the imports.  I try to remember to do it, but my formatter will sometimes remove them.

mykola-mokhnach commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,564 @@
+/**
+ * @type {`${typeof DRIVER_TYPE}s`}
+ */
+const CONFIG_DATA_DRIVER_KEY = `${DRIVER_TYPE}s`;
```
what is the magic `s` suffix?

boneskull replied:
Well... it's an `s`.  `driver` becomes `drivers`.  This is a key in the root of `extensions.yaml`.

mykola-mokhnach replied:
maybe a small explanation comment would make it cleaner. For me this magic char looked quite cryptic

mykola-mokhnach commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,564 @@
+              pkg,
+              path.join(filepath, 'package.json'),
+            );
+            didChange = true;
```
can we return right here?

mykola-mokhnach commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,564 @@
+ *
+ * It's 4 because smaller numbers didn't work.
+ */
+const DEFAULT_SEARCH_DEPTH = 4;
```
why not 5 or 6?

boneskull replied:
Because 4 was large enough :D

mykola-mokhnach commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,564 @@
+   * @returns {Promise<ManifestData>} The data
+   */
+  async read () {
+    if (this._reading) {
```
I prefer to use bluebird promises. they have convenience methods like http://bluebirdjs.com/docs/api/ispending.html, which helps to avoid nullifying the value

boneskull replied:
I don't really need anything bluebird provides here and am not using the global `Promise`...  Anyway, if we want to enforce an explicit cast of all promises to bluebird, we should put that in our eslint config.

mykola-mokhnach replied:
I usually don't really care what type of promise it is. If we need special bluebird features we could always safely call B.resolve(...) on the initial promise and then safely use its features.

mykola-mokhnach commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,564 @@
+        const yaml = await fs.readFile(this._manifestPath, 'utf8');
+        data = YAML.parse(yaml);
+      } catch (err) {
+        if (err.code === 'ENOENT') {
```
I would rather prefer to explicitly check if the file exists first. Exception-driven flow is not very obvious

boneskull replied:
I learned the exact opposite.  checking if a file exists first is more I/O overhead

mykola-mokhnach commented on packages/appium/lib/extension/package-changed.js:
```diff
@@ -0,0 +1,72 @@
+ */
+export async function packageDidChange (appiumHome) {
+  if (!appiumHome || !_.isString(appiumHome)) {
+    throw new TypeError('appiumHome must be a string');
```
such error might be confusing for a user if they see it.

boneskull replied:
if a user calls `packageDidChange` they have other problems 😄  

this exception is really for a contributor to appium who may call the function with something weird.

I could also just remove the conditional/exception and trust people will mind the types.

mykola-mokhnach replied:
if this is intended to be used by devs only then I assume they would be fine with default exception they'd get if a wrong method is called on appiumHome instance. The docs clearly states the var is expected to be a string

boneskull replied:
you're right, I'm not sure why I did that.

mykola-mokhnach commented on packages/appium/lib/extension/package-changed.js:
```diff
@@ -0,0 +1,72 @@
+  // first mkdirp the target dir.
+  const hashFilenameDir = path.dirname(hashFilename);
+  try {
+    log.debug(`Creating hash file directory: ${hashFilenameDir}`);
```
I usually try to reduce the scope of try/catch blocks as much as possible to not catch errors which we don't want to catch.

boneskull replied:
hm, yes, I can do that here

mykola-mokhnach commented on packages/appium/lib/extension/package-changed.js:
```diff
@@ -0,0 +1,72 @@
+      cwd: appiumHome,
+      hashFilename: HASHFILE_RELATIVE_PATH,
+    });
+    isChanged = result.isChanged;
```
this could be simplified to
```
({isChanged, writeHash, oldHash, hash} = result);
```

boneskull replied:
ahh, that syntax had eluded me.  parentheses!! who knew

mykola-mokhnach commented on packages/appium/lib/extension/package-changed.js:
```diff
@@ -0,0 +1,72 @@
+    oldHash = result.oldHash;
+    hash = result.hash;
+  } catch {
+    return true;
```
should we return true if an exception has happened?

boneskull replied:
This case is probably rare, but I think yes. A `true` return value causes appium to look for extensions.  If we're here, we've already determined that we're in a dir with a `package.json` and that `package.json` contains appium.  if we don't return `true` we may miss an extension that the user added.  I think it's better to do (potentially) extra work in this case.

mykola-mokhnach commented on packages/appium/lib/extension/package-changed.js:
```diff
@@ -0,0 +1,72 @@
+
+  if (isChanged) {
+    try {
+      writeHash();
```
is this call synchronous?

boneskull replied:
yes

mykola-mokhnach commented on packages/appium/lib/extension/plugin-config.js:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  /**
+   * Creates or gets an instance of {@link PluginConfig} based value of `appiumHome`
```
hm, not sure this method creates anything, but IMHO it makes sense to combine it with `create` one

boneskull replied:
yes, this is confusing and one of these functions should probably go away.

mykola-mokhnach commented on packages/appium/lib/schema/schema.js:
```diff
@@ -4,11 +4,40 @@ import Ajv from 'ajv';
+   * @param {K} key
+   * @param {V} value
+   */
+  set (key, value) {
```
IMHO it's confusing to have `set` in an immutable map. I'd rather prefer an implementation where a map can only be initialised upon creation like in https://immutable-js.com/#javaScript-first-api

boneskull replied:
I think this is a misnomer. I probably need to rename it.  It's _not_ an immutable Map.  It's a readable map which only supports "inserts"; no delete and no update.    I don't really know what to call that.

boneskull replied:
I've renamed it 😄

mykola-mokhnach commented on packages/appium/lib/schema/schema.js:
```diff
@@ -113,7 +144,7 @@ class AppiumSchema {
    */
   hasRegisteredSchema (extType, extName) {
-    return this._registeredSchemas[extType].has(extName);
+    return Boolean(this._registeredSchemas[extType].has(extName));
```
isn't `has` already returning a boolean value?

boneskull replied:
Yes, good catch.

mykola-mokhnach commented on packages/support/lib/env.js:
```diff
@@ -0,0 +1,156 @@
+  /**
+   * @param {string} [cwd] - Current working directory.  _Must_ be absolute, if specified.
+   */
+  async (cwd) => {
```
I prefer to use non anonymous functions there, because it looks more clean in the exception stack trace

boneskull replied:
probably a good idea

mykola-mokhnach commented on packages/support/lib/env.js:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    try {
+      cwd = cwd ?? process.cwd();
```
not sure about javascript, but in Java I always try to [avoid](https://softwareengineering.stackexchange.com/questions/245767/is-modifying-an-incoming-parameter-an-antipattern) reassigning of arg values and create a new variable instead

boneskull replied:
I hear what you're saying, but I'm not really convinced by many of the arguments in that blog post.  L120 is no different than assigning a default value in the function signature.  Which.. I think I can do here, so I'm going to do that.

mykola-mokhnach commented on packages/support/lib/fs.js:
```diff
@@ -23,13 +22,13 @@ const findRootCached = _.memoize(findRoot);
   async hasAccess (path) {
     try {
-      await this.access(path, _fs.R_OK);
+      await fs.access(path, _fs.R_OK);
```
is `fs` really known at this stage?

boneskull replied:
yes.

mykola-mokhnach commented on packages/support/lib/util.js:
```diff
@@ -288,16 +291,14 @@ async function isSameDestination (path1, path2, ...pathN) {
  * @throws {Error} if strict mode is enabled and `ver` cannot be coerced
  */
-function coerceVersion (ver, strict = true) {
+function coerceVersion (ver) {
```
this is a breaking change, which might render other depending code to fail. Are you sure you want to keep it?

boneskull replied:
no, there is no other code calling `coerceVersion` afaik.

mykola-mokhnach replied:
There is. A simple search through the appium org looks like

<img width="791" alt="coerce" src="https://user-images.githubusercontent.com/7767781/154356383-0b153bf8-961c-4931-bb9a-7d663e8c1cd6.png">

boneskull replied:
oh... ugh.

boneskull replied:
I think I thought this was a different, non-public utils...

boneskull replied:
Anyway, I've reverted this and fixed the type.

boneskull commented on packages/appium/lib/config.js:
```diff
@@ -296,10 +295,12 @@ async function validateTmpDir (tmpDir) {
   }
 }
 
+const rootDir = fs.findRoot(__dirname);
```
Yes. It's re-exported from the `find-root` module.

boneskull commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,215 @@
+   * @type {Record<string,DriverConfig>}
+   * @private
+   */
+  static _instances = {};
```
we could! this is mostly a stylistic choice, but I'm exposing it here for debugging/testing purposes.  I know you've mentioned you don't love `static`, but I think it's a fine place to put it. 😄

boneskull commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,215 @@
+   * @param {ManifestDriverData} extData
+   * @returns {string}
+   */
+  extensionDesc (driverName, {version, automationName}) {
```
this is old code and only appears here since I moved it.

boneskull commented:
status: 

- changing logic to search in parent dirs for a `package.json` containing an `appium` dependency. if I don't do this, running appium from any subdirectory of a working copy containing a `package.json` depending on `appium` will likely result in using the _default_ `APPIUM_HOME` instead of the parent dir of `package.json` like it should. 
- e2e tests

I'm finding the line between what goes in `@appium/support` and what doesn't kind of fuzzy. questions:

- should the `env` module actually live in `@appium/support`? I'm thinking yes, because other packages or scripts may have need for resolving `APPIUM_HOME`.  
- do all of the functions in `env` actually belong there? for instance, the functions which deal with reading/finding `package.json`.
- likewise, should `package-changed` (currently in `appium/lib/extension`) move somewhere else?  it can trivially be made generic

@jlipps @mykola-mokhnach

jlipps commented:
should the env module actually live in @appium/support?

yeah, putting some of those things in `@appium/support` if we think other packages will use them isn't a bad idea.

> do all of the functions in env actually belong there?

seems ok

> likewise, should package-changed (currently in appium/lib/extension) move somewhere else? it can trivially be made generic

up to you

boneskull committed:
feat(support): move npm module into support

- Altered strategy for determining `APPIUM_HOME` by using `npm ls` and reading `package.json` if that fails
- rename `isLocalAppiumInstalled` to `hasAppiumDependency` in `env` module
- no longer care about `peerDependencies`; we can add this later if necessary
- consume `pkg-dir` module
- remove `read-pkg-up`
- move `npm` module into `support` (which is the `feat` here, I guess)
- the `NPM` class no longer has internal state, so doesn't really need to be a class, but I didn't want to bother with that refactor.  the module exports the class `NPM` and an instance `npm`.  creating new instances of `NPM` are useless, but there's no guard against it.
- method signatures in `NPM` class have changed to generally accept a `cwd` string.  further, non-optional options have been moved out of an optional options object. 😄

feat(appium): remove NPM module (it went into @appium/support)

- Probably not a breaking change, but it certainly breaks a few things in `appium`, which I addressed.
- Updated for new `NPM` api
- fixed sandboxing issue in `cli-specs`
- update `mocks`

boneskull commented:
that failure is a good head-scratcher since I didn't change anything there.  this will be fun to debug

boneskull committed:
chore(gulp-plugins): replace find-root with pkg-dir

`find-root` is unmaintained

boneskull committed:
feat(support): remove mkdirp

- `mkdirp` module is now deprecated and re-exports `fs.mkdirp`, which is new
- fix some types in `fs` module
- fix `walkDir` which did not actually resolve to `null` ever

boneskull committed:
chore(appium): use new mkdirp

- fix some types, mocks
- rename `HASHFILE_RELATIVE_PATH` to `PKG_HASHFILE_RELATIVE_PATH`; hopefully that's less opaque
- fixed broken junk in `commands-yml/parse.js` which will conflict with #16459 so idk

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16443** <a href="https://app.graphite.dev/github/pr/appium/appium/16443" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  (this PR)
    * **PR #16482** <a href="https://app.graphite.dev/github/pr/appium/appium/16482" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #16459** <a href="https://app.graphite.dev/github/pr/appium/appium/16459" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

boneskull commented:
@jlipps this is now passing the build, so ready for your perusal.

I'll do a proper squash when it looks ok.

boneskull committed:
test(appium): add e2e tests for CLI argument errors (#16496)

This addition was prompted by #16076, which I was loathe to merge until I got this work done.

Further:

- changed the signature of `runAppiumRaw` helper to require options (because currying)
- in lieu of an actual plugin for mocha that handles snapshots, added some helper functions to help with comparing `appium`'s output
- revert all of the removals of `@ts-expect-error` I did somewhere else.  this was due to a stale TS lang server

boneskull committed:
chore(support): replace promisified fs functions with native fsPromises functions

(where applicable)

- typed and documented some stuff
- added `fs.openFile` which resolves w/a `fsPromises.FileHandle` object.  `fs.open` still resolves with a file descriptor.  this is, I think, the only explicit conflict between the builtin `fs` module and ours. we should consider deprecating or removing our `fs.open` and using the `FileHandle`-returning impl, because it's a little easier to use.  Once we do, we can also remove `fs.close()`, `fs.read()` and `fs.write()`, as those are redundant since they are methods of the `FileHandle` object returned from `fsPromises.open`.  A cursory check in the `appium` org suggests that `open` and its ilk are not used except in some test cases.
- assign properties directly to the `fs` object, as this is easier for the type system to understand (rather than iterating over property names and assigning properties in the loop)
- `which` actually returns a `Promise` now and does not need promisification
- added note about future changes to `rmdir`

jlipps commented:
ok i should be ready to review this, shall i wait for all the depended PRs to be finalized in case that triggers changes here?

boneskull commented:
@jlipps, no, those PRs are deps on _this_ one.  so once this is merged, those directly targeting this one will be retargeted to 2.0.

boneskull committed:
chore: remove gulp from appium package (#16459)

This change removes `gulp` and replaces it with plain ol' `babel` and `mocha`.

## `appium`

- No gulp.
- `test/` has been re-organized.  All test files now end in the extension `.spec.js`, which is a recognized convention and understood by editors/IDEs.  The tests are further split into `unit` and `e2e` subdirs.  This makes working with both `babel` and `mocha` easier.
- The tests are _not_ run against the transpiled code; code is now transpiled on-the-fly via `@babel/register`. This only affects `appium`.
- `commands-yml/validator.js` moved to `scripts/parse-yml-commands.js`. It has been rewritten as a CJS module.  `commands-yml/` should not contain `.js` files.
- `test/setup.js` is a new test harness specific to Appium, which loads & configures `@babel/register`.  Eventually, the contents of this file can be moved into the root monorepo's `test/setup.js`, but the rest of the packages do not need `@babel/register`, so we avoid the overhead.

## `@appium/gulp-plugins`

- Unfortunately, due to the new `npm test` strategy (see below), we cannot use the `nyan` reporter.  Or rather, we _can_, but will not see the animation.  So the reporter is now set to default to `spec`.
- Modified tests so they didn't overwrite the actual `build` dir with transpiled fixtures, which is what was happening. This was causing failures depending on the transpile/test order.

## All Packages

- All packages now have their own `build`, `test`, `test:e2e` and `dev` (just build + watch mode) npm scripts. This makes running a complete build easier, since each package can now provide its own appropriate command.  This is _mostly_ just `gulp once` (where `gulp` was used before), except for `appium`, where `@babel/cli` is used directly.
- This is what happens when `npm test` is run from the monorepo (in order):
    1. All packages are built (in parallel) by running `npm run build` in each folder (via `lerna`). See `pretest`
    2. All code is linted directly via `eslint`.
    3. All _unit tests_ are run w/ `npm test` in each folder, having assumed transpilation already occurred (if necessary).
- `npm run e2e-test` runs the `npm run test:e2e` in each folder. No transpilation occurs! Careful.
- All scripts are expected to live in `packages/*/scripts`, and this directory enforces CJS via ESLint.  `postinstall.js` moved, as well as `check-npm-pack-files.js`. `generate-schema-declarations.mjs` moved from the root `scripts/` dir into `packages/appium/scripts/`; it's now CJS instead of ESM.
- I had profiled running various things with `lerna run --parallel`, but it turns out this is not faster.  I'm not sure why, but I think it may have something to do with streaming STDOUT/STDERR, or potentially the overhead of running child processes outweighs the time it takes to actually run tests and builds.
- The behavior of `lerna run` now mimicks `lerna exec`, which was "run stuff in serial, streaming the output". The default behavior of `lerna run` is "run stuff in parallel, buffering the output".  Buffered output sucks, and streaming the output from all tasks at once creates chaos.  To override this behavior, you'll see `--parallel --concurrency=8 --prefix` for tasks that can truly be parallelized, like transpilation.  The number 8 is the "maximum", and CI usually won't have more than 2 cores available anyway.
- When run in CI, Mocha runs with `--forbid-only`, which will fail if there's an `.only()` somewhere in the tests. Further, it forces color output.

## Etc.

- Updated Wallaby config, `.gitignore`, `.eslintignore`.
- Sorted `package.json` files
- Removed `watch` and `coverage` scripts which were broken anyway.

chore(appium): fix schema json path issues

- the path to the JSON schema is now correct: `lib/appium-config-schema.json`
- add `log-symbols` to devdeps for niceness. this is already a transitive dep

boneskull committed:
chore: ban direct usage of sinon (#16485)

* chore: ban direct usage of sinon

Calling `sinon.stub()`, `sinon.mock()` etc., directly is error-prone. Care must be taken to explicitly restore any stub created.

Using a sandbox (via `sinon.createSandbox()`), you don't need to remember to restore everything you fiddled with.

To that end, I've converted all direct `sinon` calls into sandbox calls and **BANNED** direct usage of `sinon` via ESLint.

# Conflicts:
#	.eslintrc

* test(base-driver): replace restricted sinon.spy() usage in server-specs.js

* test(support): formatting

jlipps commented on packages/appium/test/setup.js:
```diff
@@ -0,0 +1,6 @@
+// This configures @babel/register to look for a babel config in parent dir(s) of
```
helpful comment 👍

boneskull replied:
yes, my verbosity increases with the size of the headache that the problem gave me

jlipps reviewed (commented):
uhhh that took a long time. i think there were one or two minor changes i requested and a few clarifying questions. overall looks good.

boneskull commented on package.json:
```diff
@@ -124,6 +130,7 @@
     "sync-monorepo-packages": "0.3.4",
     "through2": "4.0.2",
     "type-fest": "2.8.0",
+    "typescript": "4.6.0-dev.20220210",
```
yeah, last time I checked, "it doesn't work".  I'll try again though with 4.5.x

jlipps replied:
ok. should we add a comment in here like:

```
"devDependencies": {
  ...,
  "//": {
    "typescript": "4.5.x doesn't work because reasons"
  }
}
```

boneskull commented on packages/appium/lib/cli/args.js:
```diff
@@ -87,9 +82,9 @@ function makeInstallArgs (type) {
       required: false,
       default: null,
-      choices: INSTALL_TYPES,
+      choices: INSTALL_TYPES_ARRAY,
```
yeah. I thought I added a comment to that effect somewhere. `argparse` does naughty things.

boneskull commented on packages/appium/lib/cli/extension-command.js:
```diff
@@ -1,47 +1,47 @@
+
+    /** @type {ExtType} */
     this.type = type;
+    /**
```
oh, uh, no reason.  I guess we could actually document these variables

boneskull commented on packages/appium/lib/cli/extension-command.js:
```diff
@@ -121,7 +121,7 @@ export default class ExtensionCommand {
       }
       const installTxt = installed ?
-        `@${version.yellow} ${('[installed ' + typeTxt + ']').green}` :
+        `@${/** @type {string} */(version).yellow} ${('[installed ' + typeTxt + ']').green}` :
```
This isn't _truly_ a JS-in-TS thing.  `version` may be `undefined`, per the type we've given it.  the equivalent in TS would be

```ts
`@${(<string>version).yellow}`
```

...so it's still bothersome.   Anyway, whenever we see this sort of inline typing, it's a coercion.  It means one of two things: 

1. TS simply cannot understand the type.
2. The type is wrong

For example, here's a situation where TS just doesn't get it:

```ts
const foo: Set<string> = new Set([' bar ']);
if (foo.has('bar')) {
  const bar: string = foo.get('bar'); // <-- ERROR, `bar` may be `undefined`
}
```

In TS, you need to use type `string!` (iirc; the `!` means "not undefined" I think).  but it's not possible to do this through type inference alone; you have to tell the compiler that it _really_ is defined.  likewise with JS... but the only tool we have is inline types, which is more verbose but also easier to understand.

Ideally, these inline types should be kept to a bare minimum.  In this case, I think `version` _may always be a `string`_,  so I'll need to double check.

IMO the effort required to migrate to TS is not going to be worth it.

boneskull replied:
OK, the `version` _can_ be empty if the item was pulled out of `knownExtensions`.  The type is not _wrong_ per se, but it's not specific enough; if `installed` is `true`, this _implies_ that `version` is a `string`.  So I'll see if I can work that out.

jlipps replied:
I was making a simpler point, which is that ``@${/** @type {string} */(version).yellow}` is much uglier and harder to read in the code than `@${(version as string).yellow}`.

boneskull commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,230 @@
+   static create (manifest, {extData, logFn} = {}) {
+     const instance = new DriverConfig(manifest, {logFn, extData});
+     if (DriverConfig.getInstance(manifest)) {
+       throw new Error(`Manifest with APPIUM_HOME ${manifest.appiumHome} already has a DriverConfig; use DriverConfig.getInstance() to retrieve it.`);
```
I'd say yes, because the "someone" is likely a contributor/developer who did a thing wrong. 

I revisited this code a couple times.  The problem is the options object in second parameter.  A developer could be tempted to call `DriverConfig.create('/root', myOpts)` and then `DriverConfig.create('/root', otherOpts)` for some reason.  if the code just returned the `Manifest` for `/root`, what of `otherOpts`?  This would be unexpected behavior; it'd return a `Manifest` having opts `myOpts`.  

That's why there's a `create()` _and_ a `getInstance()`.

jlipps replied:
👍

boneskull commented on packages/appium/lib/extension/extension-config.js:
```diff
@@ -0,0 +1,459 @@
+   * @returns {void}
+   */
+  // eslint-disable-next-line no-unused-vars
+  print (activeNames) {
```
yes.

boneskull commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,570 @@
+    const walkOpts = _.defaults({depthLimit}, DEFAULT_FIND_EXTENSIONS_OPTS);
+    // this could be parallelized, but we can't use fs.walk as an async iterator
+    let didChange = false;
+    for await (const {stats, path: filepath} of fs.walk(
```
is this an interview question? 😄   this is my best guess; I've never seen the second syntax, but it looks valid.

```js
for await (x of y) { // y is an async iterator object. y is _not_ treated like a `Promise`
  // the promise is here. the loop does not continue until it resolves
}
```

```js
for (x of await y) {  // y may or may not be a promise, but the loop doesn't begin until it resolves to a synchronous iterator.
  // no promise in here
}
```

jlipps replied:
ok. i think you thought i was being smarter than i am. but your comment helped me to see the answer is "async iterators".

i've only ever used the second syntax. i guess i thought we needed it because it was a promise that resolved to an array. but instead it is apparently an async iterator. so we want the first syntax.

boneskull commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,570 @@
+        };
+      }
+    } else {
+      throw new TypeError(
```
This function (`addExtensionFromPackage`) is called by `syncWithInstalledExtensions`, which is called by `read()` under certain conditions.  `read()` is called by `init()`, which is called by `main()`, and this `TypeError` is going to bubble up to `main`'s rejection handler.

I'd say the people likely to encounter this error are extension developers. They should know if their extension is invalid. For that reason, I probably wouldn't want to just warn; it'd be too easy to miss the fact that their extension isn't loaded, and this would be a confusing debugging experience.

jlipps replied:
👍

boneskull commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -0,0 +1,570 @@
+
+    if (isDriver(pkgJson)) {
+      if (!this._data.drivers[pkgJson.appium.driverName]) {
+        this._data.drivers[pkgJson.appium.driverName] = {
```
good catch

boneskull commented on packages/appium/lib/extension/plugin-config.js:
```diff
@@ -0,0 +1,111 @@
+  /**
+   * A mapping of {@link Manifest} instances to {@link PluginConfig} instances.
+   *
+   * `Manifest` and {@link ExtensionConfig} have a one-to-many relationship; each `Manifest` should be associated with a `DriverConfig` and a `PluginConfig`; no more, no less.
```
It's...well, that depends if you would define "two" as "many".

jlipps replied:
it just made it hard for me to grok this comment, as if there was more structure possible here than there is. it would have read more straightforward to me as something like just: "Each `Manifest` has two  `ExtensionConfig`s; one for plugins and one for drivers"

boneskull commented on packages/appium/lib/grid-register.js:
```diff
@@ -12,9 +12,9 @@ const hubUri = (config) => {
- * @param {string} addr - Bind to this address
- * @param {number} port - Bind to this port
- * @param {string} basePath - Base path for the grid
+ * @param {string} [addr] - Bind to this address
```
Yes.  If `data` is a `string`, they are required. If `data` is an `object`, they are not.  I'll make sure there's a comment to that effect.

boneskull commented on packages/appium/lib/schema/schema.js:
```diff
@@ -4,12 +4,42 @@ import Ajv from 'ajv';
+ * @template K,V
+ * @extends {Map<K,V>}
+ */
+export class RoachHotelMap extends Map {
```
I renamed `ReadonlyMap` to `RoachHotelMap` because @mykola-mokhnach pointed out that it was a misnomer.  It's _not_ "read-only", since you can write to it.  But you _can't_ take things out or change what's already there.

boneskull commented on packages/appium/lib/extension/driver-config.js:
```diff
@@ -0,0 +1,230 @@
+       log.info(`Will require it at ${this.getInstallPath(driverName)}`);
+       const driver = this.require(driverName);
+       if (!driver) {
+         throw new Error(`MainClass ${mainClass} did not result in a driver object`);
```
err, actually, it's not instantiating the class, it's just returning the class/constructor itself. but I can make it better

boneskull committed:
fix(appium): remove bad log

the log message upon "appium driver install foo" said the strategy was "null", which is not quite right.

boneskull committed:
test(base-driver): rewrite some assertions in capability-specs

simplified some stuff that was giving me problems.

instead of doing a try..await..catch and calling `fail()` if the catch block wasn't called, just use `sinon-chai`'s `.should.eventually.be.rejectedWith(ErrorClass, /message/)`

depending on the order in which this test file is run versus others, it's possible that some will fail because of global logging weirdness.  but normally this doesn't happen... keep an eye on it!

boneskull committed:
chore: use proper version of typescript

boneskull committed:
chore: make typedefs less verbose

`@typedef {Object} Foo` is the same as `@typedef Foo`, and one of those is shorter.

boneskull committed:
chore(appium): peer review changes

- add note about argparse tomfoolery
- refactor `list` command impl for better types
- rewrite exception in `driver-config`
- add `hasDriver`, `hasPlugin`, and small refactors in `manifest`
- add note about args in `grid-register`
- remove invalid comment in `main`
- more test coverage for `manifest`

boneskull commented:
I am the king of git and I am going to merge this now

jlipps commented:
\o/ woooo!!!
